### PR TITLE
DBZ-3028 Validate field.exclude.list & field.renames for MongoDB connector

### PR DESCRIPTION
**Pull request by ani-sha**
https://issues.redhat.com/browse/DBZ-3028

### CODE REVIEW

File: debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/MongoDbConnectorIT.java [hunk 1]
```diff
@@ -143,6 +144,44 @@ public void shouldFailToValidateInvalidConfiguration() {
...
+
+    }
+
+    @Test
+    public void shouldNotThrowErrorWithValidFilterConfiguration() {
+        shouldValidateFilterFieldConfiguration(MongoDbConnectorConfig.FIELD_EXCLUDE_LIST, "db1.collectionA.name", 0);
+        shouldValidateFilterFieldConfiguration(MongoDbConnectorConfig.FIELD_EXCLUDE_LIST, "db1.collectionA.name,db2.collectionB.house", 0);
+
+        shouldValidateFilterFieldConfiguration(MongoDbConnectorConfig.FIELD_RENAMES, "db1.collectionA.name:new_name", 0);
+        shouldValidateFilterFieldConfiguration(MongoDbConnectorConfig.FIELD_RENAMES, "db1.collectionA.name:new_name,db2.collectionB.house:new_house", 0);
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/MongoDbConnectorIT.java:166]: Can we add tests that include additional whitespace characters, e.g. at the beginning or the end, and around the comma (", ")?

File: debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/FiltersTest.java [hunk 2]
```diff
@@ -162,41 +161,6 @@ public void shouldExcludeCollectionCoveredByLiteralInBlacklist() {
         assertCollectionIncluded("db2.collectionA");
     }
 
-    @Test(expected = ConfigException.class)
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/FiltersTest.java:165]: FWIW, I think this approach using separate test methods was advantageous: if one fails, you know exactly which one it is; whereas now, you'll have to refer to the line number and see which one a failed assertion is about. Also, if there are multiple failures, all the affected test methods would have failed before, whereas now with the larger test methods, one failure may "hide" another failure, i.e. the latter will only become visible once the former has been fixed.

File: debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/MongoDbConnectorIT.java [hunk 3]
```diff
@@ -143,6 +144,44 @@ public void shouldFailToValidateInvalidConfiguration() {
         assertNoConfigurationErrors(result, CommonConnectorConfig.TOMBSTONES_ON_DELETE);
     }
 
+    @Test
+    public void shouldThrowErrorWithInvalidFilterConfiguration() {
+        shouldValidateFilterFieldConfiguration(MongoDbConnectorConfig.FIELD_EXCLUDE_LIST, ".name", 1);
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/MongoDbConnectorIT.java:149]: Let's also add the value from the original Jira report, "inventory".

**Review by gunnarmorling — CHANGES_REQUESTED** [review]
Some comments/suggestions inline, @ani-sha.

File: debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/FiltersTest.java (hunk 2 above)
Comment (rk3rn3r) [reply debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/FiltersTest.java:165]: You could provide custom failure messages too to recognize the failing test.

Comment (gunnarmorling) [reply debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/FiltersTest.java:165]: This won't prevent the "hiding" issue, though.

Comment (gunnarmorling) [reply debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/FiltersTest.java:165]: Also, you cannot re-run individual tests/assertions e.g. in your IDE. In general, finer-grained test methods are preferable for all these reasons.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java [hunk 4]
```diff
@@ -39,6 +40,13 @@
     protected static final String DATABASE_INCLUDE_LIST_ALREADY_SPECIFIED_ERROR_MSG = "\"database.include.list\" is already specified";
     protected static final String DATABASE_WHITELIST_ALREADY_SPECIFIED_ERROR_MSG = "\"database.whitelist\" is already specified";
 
+    protected static final Pattern PATTERN_SPILT = Pattern.compile(",");
+
+    protected static final Pattern FIELD_EXCLUDE_LIST_PATTERN = Pattern.compile("^(?![.])([*\\w]+.[\\w]+.[\\w]+)+(.[\\w]+)*$");
+    protected static final String QUALIFIED_FIELD_EXCLUDE_LIST_PATTERN = "<databaseName>.<collectionName>.<fieldName>.<nestedFieldName>";
```
Comment (ani-sha) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java:46]: Can these regexes get simpler? As I checked it currently accepts `inventory` or `inventory.name` values which it shouldn't. Any suggestions for modification?

**Comment by ani-sha** [comment]
@gunnarmorling Expanded the test cases and addressed the review comments. Ready for another review.

File: debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/MongoDbConnectorIT.java [hunk 5]
```diff
@@ -143,6 +144,96 @@ public void shouldFailToValidateInvalidConfiguration() {
...
+    }
+
+    @Test
+    public void shouldThrowExceptionWhenFieldExcludeListFieldPartIsMissing() {
+        shouldValidateFilterFieldConfiguration(MongoDbConnectorConfig.FIELD_EXCLUDE_LIST, "db1.collectionA.", 1);
+    }
+
+    @Test
+    public void shouldNotThrowExceptionWhenFieldExcludeListHasLeadingWhiteSpaces() {
+        shouldValidateFilterFieldConfiguration(MongoDbConnectorConfig.FIELD_EXCLUDE_LIST, " *.collectionA.name", 0);
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/MongoDbConnectorIT.java:169]: On all the "should not throw" methods, do we have corresponding tests in `FiltersTest` which assert that the filters actually work as intended, by asserting against matching and non matching field names? In particular, I'm wondering about the whitespace behavior: is this handled correctly both during validation and actual filter processing?

**Review by gunnarmorling — COMMENTED** [review]
Thx, @ani-sha! This looks mostly good to me now. One more question inline, which may lead to some more work in the actual filter implementation and corresponding tests.

File: debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/MongoDbConnectorIT.java (hunk 5 above)
Comment (ani-sha) [reply debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/MongoDbConnectorIT.java:169]: @gunnarmorling I have added the respective test cases for all the "should not throw" methods in `FiltersTest` by doing a `NotNull` check for now. So to exactly match and assert how may I get the values out of the `FieldSelector`?

Comment (gunnarmorling) [reply debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/MongoDbConnectorIT.java:169]: Something like this should work:

```
        filters = build.excludeFields(" *.collectionA.key1").createFilters();
        assertEquals(
        		Document.parse(" { \"key1\" : \"value1\", \"key2\" : \"value2\" }"),
        		filters.fieldFilterFor(null).apply(Document.parse(" { \"key2\" : \"value2\" }"))
        );
```

Also the test names should be adjusted accordingly, i.e. not focusing on the fact of not throwing an exception but e.g. "excludeFilterShouldRemoveMatchingField()", etc.

Comment (ani-sha) [reply debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/MongoDbConnectorIT.java:169]: Thanks, @gunnarmorling, I have added the tests accordingly and adjusted the test names as well.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/FieldSelector.java [hunk 6]
```diff
@@ -116,12 +116,12 @@ public FieldSelector build() {
 
         private void parse(String value, Function<String, Path> pathCreator, List<Path> result) {
             if (!Strings.isNullOrEmpty(value)) {
-                Arrays.stream(value.split(",")).forEach(name -> result.add(pathCreator.apply(name)));
+                Arrays.stream(value.trim().split(",")).forEach(name -> result.add(pathCreator.apply(name)));
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/FieldSelector.java:119]: Very nice; great to see this change, as I was expecting the new test to run into this. Thanks!

**Review by gunnarmorling — APPROVED** [review]

**Comment by gunnarmorling** [comment]
Applied; thx, @ani-sha!

**Comment by rk3rn3r** [comment]
Nice! Thx a lot @ani-sha !
